### PR TITLE
fix(trace): properly report `fallback()` fn

### LINKED ISSUE #9115: bug: `--gas-report` groups all methods of all contracts that use the same type of proxy contract
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

_No response_

### What command(s) is the bug in?

_No response_

### Operating System

macOS (Apple Silicon)

### Describe the bug

After trying to generate `--json` output where i faced https://github.com/foundry-rs/foundry/issues/9111 I played a bit around and faced another "bug"? in regards to the output in general.

If you have multiple contracts in a repo using the same type of proxy (e.g. oz `TransparentUpgradeableProxy`), foundy will group all methods of all contracts using this proxy together. What is even worse, i think when multiple implementations implement the same signature (e.g. `transfer` - it will show as one and consider all implementations for things like average etc).

---

Not 100% sure what the solution is, perhaps it could be possible to differentiate multiple Proxies dependent on the impl being delegate-called?

### PR DESCRIPTION
**Pull request by grandizzy**
## Motivation
closes #9115 

## Solution
- when decoding functions for contracts with fallback make sure the decoded function is in contract sigs, otherwise use fallback for trace

### CODE REVIEW

File: crates/forge/tests/cli/cmd.rs [hunk 1]
```diff
@@ -2818,3 +2818,83 @@ forgetest_init!(gas_report_include_tests, |prj, cmd| {
...
+
+| test/DelegateProxyTest.sol:ProxiedContract contract |                 |      |        |      |         |
+|-----------------------------------------------------|-----------------|------|--------|------|---------|
+| Deployment Cost                                     | Deployment Size |      |        |      |         |
+| 106511                                              | 276             |      |        |      |         |
+| Function Name                                       | min             | avg  | median | max  | # calls |
+| deposit                                             | 3320            | 3320 | 3320   | 3320 | 1       |
+...
+
+"#]]);
```
Comment (zerosnacks) [inline crates/forge/tests/cli/cmd.rs:2899]: Would you mind adding an additional JSON test for completeness and moving it up to below this block?

https://github.com/foundry-rs/foundry/blob/a40b59c4886ab5da9995b96ccd97bf821f12e7b6/crates/forge/tests/cli/cmd.rs#L2392

**Review by zerosnacks — COMMENTED** [review]
Nice! Small note in regards to test case completeness w/ JSON

@sakulstra would you mind double checking the proposed output format?

File: crates/forge/tests/cli/cmd.rs (hunk 1 above)
Comment (grandizzy) [reply crates/forge/tests/cli/cmd.rs]: yep, added and moved in https://github.com/foundry-rs/foundry/pull/9287/commits/7696273cef95ca9efdede0a015d3731728c78a02

File: crates/forge/tests/cli/cmd.rs [hunk 2]
```diff
@@ -2390,6 +2390,136 @@ contract CounterTest is DSTest {
...
+
+contract ProxiedContract {
+    uint256 public amount;
+
+    function deposit(uint256 aba) external {
+        amount = amount * 2;
+    }
+
+    function deposit() external {
+    }
```
Comment (zerosnacks) [inline crates/forge/tests/cli/cmd.rs:2408]: It the shadowing of `deposit` here intended?

Comment (grandizzy) [reply crates/forge/tests/cli/cmd.rs:2408]: yeah, to make sure that if deposit fn with different params exists on proxy contract then gas is accounted for proxy (that is we match on function signatures and not on function names).

**Review by zerosnacks — APPROVED** [review]
Lgtm, pending final check from @sakulstra

**Comment by sakulstra** [comment]
looks fine to me, thanks a lot for looking into this!

**Pull request merged**